backend/onyx/server/middleware/rate_limiting.py

```diff
@@ -0,0 +1,46 @@
+from collections.abc import Callable
+from typing import List
+
+from fastapi import Depends
+from fastapi import Request
+from fastapi_limiter import FastAPILimiter
+from fastapi_limiter.depends import RateLimiter
+from redis import asyncio as aioredis
+
+from onyx.configs.app_configs import RATE_LIMIT_MAX_REQUESTS
+from onyx.configs.app_configs import RATE_LIMIT_WINDOW_SECONDS
+from onyx.configs.app_configs import REDIS_HOST
+from onyx.configs.app_configs import REDIS_PASSWORD
+from onyx.configs.app_configs import REDIS_PORT
+
+
+async def setup_limiter() -> None:
+    redis = await aioredis.from_url(
+        f"redis://{REDIS_HOST}:{REDIS_PORT}", password=REDIS_PASSWORD
+    )
+    await FastAPILimiter.init(redis)
+
+
+async def close_limiter() -> None:
+    await FastAPILimiter.close()
+
+
+def rate_limit_key(request: Request) -> str:
+    return (
+        request.client.host if request.client else "unknown"
```

Comment:
In what cases is this unknown? Is there any worry that in majority of cases it may be unknown and we'll be blocking a ton of signups?

Reply:
(Referring to updated logic) Usually, the IP is automatically passed to the server, but if a request goes through certain proxies or services that remove IP info, we fall back to “unknown.” Similarly, if a script or device doesn't send a User-Agent header (e.g., certain bots), we default to “none,” which can lead to collisions here